### PR TITLE
feature: add script to transfer amm ownership to governance

### PR DESCRIPTION
**Pull request by cwsnt**

### CODE REVIEW

**Review by tjcloa — CHANGES_REQUESTED** [review]

File: scripts/sip/sip_interaction.py [hunk 1]
```diff
@@ -411,3 +411,585 @@ def createProposalSIP0048():
 
     # Create Proposal
     createProposal(contracts['GovernorAdmin'], target, value, signature, data, description)
+
+# === Accepting Ownership of AMM Contracts ===
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply scripts/sip/sip_interaction.py:415]: Based on the discussion, there should be one SIP per timelock contract involved.

File: scripts/contractInteraction/amm.py [hunk 2]
```diff
@@ -342,4 +342,13 @@ def withdrawFromRBTCWrapperProxy(tokenAddress, to, amount):
     abiFile =  open('./scripts/contractInteraction/ABIs/RBTCWrapperProxy.json')
     abi = json.load(abiFile)
     wrapperProxy = Contract.from_abi("RBTCWrapperProxy", address=conf.contracts['RBTCWrapperProxy'], abi=abi, owner=conf.acct)
-    wrapperProxy.withdraw(tokenAddress, to, amount)
\ No newline at end of file
+    wrapperProxy.withdraw(tokenAddress, to, amount)
+
+def transferOwnershipAMMContractsToGovernance(contractAddress, destinationAddress):
+    print("Transferring ownership of {0} to {1}".format(contractAddress, destinationAddress))
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply scripts/contractInteraction/amm.py]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/5a4164e87b188060201ce758364676dfa05ecabd)

File: scripts/contractInteraction/contract_interaction.py [hunk 3]
```diff
@@ -107,5 +107,122 @@ def main():
 
     # getLMInfo()
 
+    # Transferring Ownership of AMM Contracts to Governance
+    # === SovrynSwapNetwork ===
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply scripts/contractInteraction/contract_interaction.py]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/5a4164e87b188060201ce758364676dfa05ecabd)

File: scripts/contractInteraction/mainnet_contracts.json [hunk 4]
```diff
@@ -30,7 +30,14 @@
     "RIF": "0x2acc95758f8b5f583470ba265eb685a8f45fc9d5",
     "MYNT": "0x2e6B1d146064613E8f521Eb3c6e65070af964EbB",
     "swapNetwork": "0x98aCE08D2b759a265ae326F010496bcD63C15afc",
+    "ammSwapSettings": "0x3aB00BEFDd7Bfc0667DE6483D2D3b2F9A74AF2da",
+    "ammContractRegistry": "0x46EBC03EF2277308BdB106a73d11c65109C4B89B",
+    "ammOracleWhitelist": "0x8e75774Ef928cE730255AB594dD1b9F0a725a56b",
     "ConverterRegistry": "0x31A0F8400c75d52FdB413372233F28E3bdFB1c06",
+    "ConverterFactory": "0xcF46f24423B8da97E2c06B41df28163D55e80935",
+    "ConversionPathFinder": "0xcF46f24423B8da97E2c06B41df28163D55e80935",
```
Comment (tjcloa) [inline scripts/contractInteraction/mainnet_contracts.json:38]: wrong address - the same as ConverterFactory

**Review by tjcloa — CHANGES_REQUESTED** [review]

**Comment by ororopickpocket** [comment]
@cwsnt please address the other issues tyrone addresses as well.

File: scripts/contractInteraction/mainnet_contracts.json (hunk 4 above)
Comment (cwsnt) [reply scripts/contractInteraction/mainnet_contracts.json]: fixed in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/0e0566114d689e3c539e0d9ee073d202c8366999)

File: scripts/sip/sip_interaction.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cwsnt) [reply scripts/sip/sip_interaction.py:415]: fixed in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/8e83e68541e0f2e82978ed1378b9864c47ffa323)

File: scripts/contractInteraction/tasks/transfer_amm_ownership_to_governance.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+def main():
+    # Transferring Ownership of AMM Contracts to Governance
+    # === SovrynSwapNetwork ===
+    print("Transfer ownership of sovrynSwapNetwork to timelockAdmin: ", conf.contracts['TimelockAdmin'])
+    transferOwnershipAMMContractsToGovernance(conf.contracts['swapNetwork'], conf.contracts['TimelockAdmin'])
+
+
+    # === SwapSettings ===
+    print("Transfer ownership of swapSettings to timelockAdmin: ", conf.contracts['TimelockAdmin'])
+    transferOwnershipAMMContractsToGovernance(conf.contracts['ammSwapSettings'], conf.contracts['TimelockAdmin'])
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply scripts/contractInteraction/tasks/transfer_amm_ownership_to_governance.py]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/8e83e68541e0f2e82978ed1378b9864c47ffa323)

**Review by ororopickpocket — APPROVED** [review]
looks good to me, thanks.

**Review by tjcloa — APPROVED** [review]
pls test the script on a forked mainnet first

**Comment by tjcloa** [comment]
pls test the script on a forked mainnet first

**Comment by tjcloa** [comment]
@cwsnt pls see the Description of this ticket https://sovryn.atlassian.net/browse/SOV-1044 to fix the governor actions overflow issue

**Review by tjcloa — CHANGES_REQUESTED** [review]
the current SIP script is not workable due to the proposals qty execution limitations
pls consider this option to fix it: https://sovryn.atlassian.net/browse/SOV-1044

**Review by ororopickpocket — APPROVED** [review]
nice job! just one tiny thing: the function names in sip_interaction are not great. please rename them to match the actual SIP numbers

**Review by tjcloa — CHANGES_REQUESTED** [review]
apart from the inline change requirements also pls resolve conflicts

File: scripts/contractInteraction/amm.py [hunk 6]
```diff
@@ -345,6 +346,28 @@ def withdrawFromRBTCWrapperProxy(tokenAddress, to, amount):
     wrapperProxy = Contract.from_abi("RBTCWrapperProxy", address=conf.contracts['RBTCWrapperProxy'], abi=abi, owner=conf.acct)
     wrapperProxy.withdraw(tokenAddress, to, amount)
 
+def transferOwnershipAMMContractsToGovernance(contractAddress, destinationAddress, contractName=''):
```
Comment (tjcloa) [inline scripts/contractInteraction/amm.py:349]:
```suggestion
def transferOwnershipAMMContractsToGovernance(contractAddress, newOwnerAddress, contractName=''):
```

File: scripts/contractInteraction/amm.py [hunk 7]
```diff
@@ -345,6 +346,28 @@ def withdrawFromRBTCWrapperProxy(tokenAddress, to, amount):
...
+
+        if(_contractAddress != contractAddress):
+            raise Exception("Unmatched contract address with the on-chain")
+
+    currentOwner = ammContract.owner()
+    if(currentOwner != conf.contracts['multisig']):
+        raise Exception("Multisig is not the owner")
+
+    print("Transferring ownership of {0} to {1}".format(contractAddress, destinationAddress))
+    data = ammContract.transferOwnership.encode_input(destinationAddress)
```
Comment (tjcloa) [inline scripts/contractInteraction/amm.py:367]:
```suggestion
    print("Transferring ownership of {0} to {1}".format(contractAddress, newOwnerAddress))
    data = ammContract.transferOwnership.encode_input(newOwnerAddress)
```

Comment (cwsnt) [reply scripts/contractInteraction/amm.py]: Updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/94459ccdbffb4c8e2fa42981a355178c076b9a2f)

File: scripts/contractInteraction/amm.py (hunk 6 above)
Comment (cwsnt) [reply scripts/contractInteraction/amm.py]: Updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/94459ccdbffb4c8e2fa42981a355178c076b9a2f)

File: hardhat/tasks/sips/args/sipArgs.js [hunk 8]
```diff
@@ -199,8 +199,152 @@ const getArgsSip0063 = async (hre) => {
...
+        values.push(0);
+        signatures.push("acceptOwnership()");
+        datas.push("0x");
+    }
+    const args = {
+        targets: targets,
+        values: values,
+        signatures: signatures,
+        data: datas,
+        description: "SIP-0067 : Accepting ownership of AMM contracts Part 2",
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply hardhat/tasks/sips/args/sipArgs.js]: Updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/94459ccdbffb4c8e2fa42981a355178c076b9a2f)

File: hardhat/tasks/sips/args/sipArgs.js [hunk 9]
```diff
@@ -199,8 +199,152 @@ const getArgsSip0063 = async (hre) => {
...
+        values.push(0);
+        signatures.push("acceptOwnership()");
+        datas.push("0x");
+    }
+    const args = {
+        targets: targets,
+        values: values,
+        signatures: signatures,
+        data: datas,
+        description: "SIP-0066 : Accepting ownership of AMM contracts Part 1",
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply hardhat/tasks/sips/args/sipArgs.js]: Updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/94459ccdbffb4c8e2fa42981a355178c076b9a2f)

File: hardhat/tasks/sips/args/sipArgs.js [hunk 10]
```diff
@@ -199,8 +199,152 @@ const getArgsSip0063 = async (hre) => {
...
+    return { args, governor: "GovernorOwner" };
+};
+
 module.exports = {
     getArgsSip0058,
     getArgsSip0049,
     getArgsSip0063,
+    getArgsSip0064,
+    getArgsSip0065,
+    getArgsSip0066,
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply hardhat/tasks/sips/args/sipArgs.js]: Updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/94459ccdbffb4c8e2fa42981a355178c076b9a2f)

**Comment by jjmr007** [comment]
I can help with the point 5

**Review by tjcloa — CHANGES_REQUESTED** [review]
Pls consider the following:   

1) amm oracles and other dependent deployments can be removed (see inline comments for details) because they can change - the actual addresses should be queried from the contract where the dependency address is stored (like oracles from converters). i suggest to start from this refactoring.  
2) all mainnet external deployments should go into `external/deployments/rskMainnet` folder (`external/deployments/rskTestnet` for the testnet) because these paths are set in hh config to look up for external deployments. no need for rskSovrynMainnet /Testnet for external deployment, please move the deployment .json data not removed at previous step to rskMainnet/rskTestnet.   respectively

**Review by tjcloa — COMMENTED** [review]
see inline comments

File: hardhat/helpers.js [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+            deploymentTarget.deployment.abi,
+            deploymentTarget.deployment.address
+        );
+        const lpToken = await ammConverterContract.token();
+
+        const ammConverterRegistryAddress = (await get("AmmConverterRegistry")).address;
+        const ammConverterRegistryContract = await ethers.getContractAt(
+            ammConverterRegistryInterface,
+            ammConverterRegistryAddress
+        );
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply hardhat/helpers.js]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/054c8482754993abd0e5d69ff0348194e7c6d845)

File: hardhat/helpers.js [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+        return registeredAddressOnchain;
+    } else if (converterType === sourceContractTypesToValidate.ConverterV2) {
+        const ammConverter = await get(converterName);
+        const ammConverterInterface = new ethers.utils.Interface(ammConverter.abi);
+
+        const ammConverterContract = await ethers.getContractAt(
+            ammConverterInterface,
+            ammConverter.address
+        );
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply hardhat/helpers.js]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/054c8482754993abd0e5d69ff0348194e7c6d845)

File: hardhat/tasks/sips/args/sipArgs.js [hunk 13]
```diff
@@ -238,9 +240,353 @@ const getArgsSip0065 = async (hre) => {
     return { args, governor: "GovernorOwner" };
 };
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply hardhat/tasks/sips/args/sipArgs.js:260]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/054c8482754993abd0e5d69ff0348194e7c6d845)

File: hardhat/tasks/sips/args/sipArgs.js [hunk 14]
```diff
@@ -238,9 +240,353 @@ const getArgsSip0065 = async (hre) => {
...
+            );
+            if (oracleAddress === ZERO_ADDRESS) return process.exit;
+            const oracleArtifact = await deployments.getArtifact("Oracle");
+            deploymentTarget.deployment = await ethers.getContractAt(
+                oracleArtifact.abi,
+                oracleAddress
+            );
+        } else {
+            const isValid = await validateAmmOnchainAddresses(deploymentTarget);
+            if (!isValid) return process.exit;
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply hardhat/tasks/sips/args/sipArgs.js]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/054c8482754993abd0e5d69ff0348194e7c6d845)

File: hardhat/tasks/sips/args/sipArgs.js [hunk 15]
```diff
@@ -238,9 +240,353 @@ const getArgsSip0065 = async (hre) => {
...
+    const signatures = [];
+    const datas = [];
+    for (let i = 0; i < deploymentTargets.length; i++) {
+        deploymentTarget = deploymentTargets[i];
+        if (deploymentTarget.deployment === "oracle") {
+            const oracleAddress = await getAmmOracleAddress(
+                deploymentTarget.sourceContractNameToValidate,
+                deploymentTarget.sourceContractTypeToValidate
+            );
+            if (oracleAddress === ZERO_ADDRESS) return process.exit;
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply hardhat/tasks/sips/args/sipArgs.js]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/054c8482754993abd0e5d69ff0348194e7c6d845)

File: hardhat/tasks/sips/args/sipArgs.js [hunk 16]
```diff
@@ -238,9 +240,353 @@ const getArgsSip0065 = async (hre) => {
...
+            sourceContractNameToValidate: "AmmContractRegistry",
+        },
+        {
+            deployment: await get("AmmSwapSettings"),
+            contractName: "SwapSettings",
+            sourceContractTypeToValidate: "ContractRegistry",
+            sourceContractNameToValidate: "AmmContractRegistry",
+        },
+        {
+            // @todo need to check the discrepancy address between onchain with the excel one
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply hardhat/tasks/sips/args/sipArgs.js]: removed in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/054c8482754993abd0e5d69ff0348194e7c6d845)

File: hardhat/tasks/sips/args/sipArgs.js [hunk 17]
```diff
@@ -238,9 +240,353 @@ const getArgsSip0065 = async (hre) => {
     return { args, governor: "GovernorOwner" };
 };
 
+const getArgsSip0067 = async (hre) => {
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply hardhat/tasks/sips/args/sipArgs.js]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/054c8482754993abd0e5d69ff0348194e7c6d845)

File: hardhat/tasks/sips/args/sipArgs.js [hunk 18]
```diff
@@ -1,5 +1,7 @@
 const { HardhatRuntimeEnvironment } = require("hardhat/types");
 const { getStakingModulesNames } = require("../../../../deployment/helpers/helpers");
+const { validateAmmOnchainAddresses, getAmmOracleAddress } = require("../../../helpers");
+const { ZERO_ADDRESS } = require("@openzeppelin/test-helpers/src/constants");
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply hardhat/tasks/sips/args/sipArgs.js]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/054c8482754993abd0e5d69ff0348194e7c6d845)

File: hardhat/helpers.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (cwsnt) [reply hardhat/helpers.js]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/450/commits/054c8482754993abd0e5d69ff0348194e7c6d845)

File: hardhat/helpers.js [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+        deploymentTarget.sourceContractTypeToValidate ===
+        sourceContractTypesToValidate.ConverterRegistry
+    ) {
+        const ammConverterRegistry = await get(deploymentTarget.sourceContractNameToValidate);
+        const ammConverterRegistryInterface = new ethers.utils.Interface(ammConverterRegistry.abi);
+
+        const ammConverterContract = await ethers.getContractAt(
+            deploymentTarget.deployment.abi,
+            deploymentTarget.deployment.address
+        );
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply hardhat/helpers.js]: updated

**Review by tjcloa — APPROVED** [review]

**Pull request merged**